README.md

|![](https://upload.wikimedia.org/wikipedia/commons/thumb/1/17/Warning.svg/156px-Warning.svg.png) | Deis Workflow will soon no longer be maintained.<br />Please [read the announcement](https://deis.com/blog/2017/deis-workflow-final-release/) for more detail. |
|---:|---|
| 09/07/2017 | Deis Workflow [v2.18][] final release before entering maintenance mode |
| 03/01/2018 | End of Workflow maintenance: critical patches no longer merged |

# stdout metrics
[![Go Report Card](http://goreportcard.com/badge/deis/stdout-metrics)](http://goreportcard.com/report/deis/stdout-metrics)
[![Build Status](https://travis-ci.org/deis/stdout-metrics.svg?branch=master)](https://travis-ci.org/deis/stdout-metrics)

Deis (pronounced DAY-iss) is an open source PaaS that makes it easy to deploy and manage
applications on your own servers. Deis builds on [Kubernetes](http://kubernetes.io/) to provide
a lightweight, [Heroku-inspired](http://heroku.com) workflow.

![Deis Graphic](https://getdeis.blob.core.windows.net/get-deis/deis-graphic-small.png)

[v2.18]: https://github.com/deisthree/workflow/releases/tag/v2.18.0


CHANGELOG.md
### v2.0.0-rc2 -> v2.0.0

#### Documentation

- [`d39fcc1`](https://github.com/deisthree/stdout-metrics/commit/d39fcc1a3d73f1b2d4b7d0382caec4b0513a4fcf) CHANGELOG.md: add entry for v2.0.0-rc2

### v2.0.0-rc1 -> v2.0.0-rc2

### v2.0.0-beta4

#### Features

 - [`3ab97e1`](https://github.com/deisthree/stdout-metrics/commit/3ab97e118f1b261ba58206752f879d1effbbcdf2) server: correctly parse router log messages from stream

#### Fixes

- [`b508ed5`](https://github.com/deisthree/stdout-metrics/commit/b508ed52c19ad8f60e64e19f9f86de8705b1681d) fix: Using wrong index when parsing router message


util/utils_test.go
package util

import (
	. "github.com/onsi/ginkgo"
	. "github.com/onsi/gomega"
)

//Nginx log format
//'[$time_iso8601] - $app_name - $remote_addr - $remote_user - $status - "$request" - $bytes_sent - "$http_referer" - "$http_user_agent" - "$server_name" - $upstream_addr - $http_host - $upstream_response_time - $request_time';
var _ = Describe("Utils", func() {
	Context("log message", func() {
		Context("with valid json", func() {
			var (
				validLogMessage  = `{"log":"[2016-05-06T20:04:23+00:00] - deis/deis-monitor-grafana - 10.240.0.15 - - - 200 - \"GET /foo HTTP/1.1\" - 31947 - \"http://grafana.mydomain.com/dashboard/db/deis-router\" - \"Mozilla/5.0 (Macintosh;\" - \"~^grafana\\x5C.(?<domain>.+)$\" - 10.167.252.252:80 - grafana.domain.com - 0.064 - 0.064\n","stream":"stderr","docker":{"container_id":"b495e59972c6d6030bc234a41951fc564dd33bead9b92915896e06abf8a8c156"},"kubernetes":{"namespace_name":"deis","pod_id":"825fbad6-13c5-11e6-9d94-42010a800021","pod_name":"deis-router-9a41n","container_name":"deis-router","labels":{"app":"deis-router"},"host":"some-valid-host"}}`
				messageJSON, err = ParseMessage(validLogMessage)
				parsedMessage, _ = ParseNginxLog(messageJSON["log"].(string))
			)

			It("should correctly parse into a map", func() {
				Expect(messageJSON).To(HaveLen(4))
			})

			It("should not error", func() {
				Expect(err).NotTo(HaveOccurred())
			})

			It("should be from deis-router", func() {
				Expect(FromContainer(messageJSON, "deis-router"), true)
			})

			It("should not be from deis-logger", func() {
				Expect(FromContainer(messageJSON, "deis-logger"), true)
			})

			It("should be able to return host", func() {
				Expect(GetHost(messageJSON)).To(Equal("some-valid-host"))
			})

			It("should parse the nginx log line", func() {
				Expect(parsedMessage).To(HaveLen(6))
			})

			It("should correctly parse app name", func() {
				Expect(parsedMessage["app"]).To(Equal("deis/deis-monitor-grafana"))
			})
		})
	})

	Context("is a plain string", func() {
		It("should return an error", func() {
			_, err := ParseMessage("invalid log message")
			Expect(err).To(HaveOccurred())
		})
	})

	Context("with invalid log structure", func() {
		It("ParseNginxLog should return an error", func() {
			invalidLogMessage := `{"log":"2016/05/04 18:37:13 INFO: Router configuration has changed in k8s.\n","stream":"stderr","docker":{"container_id":"436bec6d64b57c072b074e8704a2551a1197f5a3f438e45639a33aac3f7a615d"},"kubernetes":{"namespace_name":"deis","pod_id":"ac2ab3b8-1216-11e6-bb85-42010a800021","pod_name":"deis-router-dqgcj","container_name":"deis-router","labels":{"app":"deis-router"},"host":"gke-jchauncey-default-pool-7ae1c279-c4zl"}}`
			messageJSON, _ := ParseMessage(invalidLogMessage)
			_, err := ParseNginxLog(messageJSON["log"].(string))
			Expect(err).To(HaveOccurred())
		})
	})

	Context("malformed json", func() {
		It("should return an error", func() {
			_, err := ParseMessage(`{"log":invalid json}`)
			Expect(err).To(HaveOccurred())
		})
	})
})


influx/influx.go
package influx

import (
	"fmt"
	"os"
	"time"

	"github.com/influxdata/influxdb/client/v2"
)

//Query returns the result and an error
func Query(influxClient client.Client, queryString string) ([]client.Result, error) {
	query := client.NewQuery(queryString, "kubernetes", "s")
	response, err := influxClient.Query(query)
	if err != nil {
		return nil, err
	}

	if response.Error() != nil {
		return nil, response.Error()
	}
	return response.Results, nil
}

//WriteInt64 takes a int64 value and sends it to influx
func WriteInt64(influxClient client.Client, value int64, name string, tags map[string]string, timestamp time.Time) error {
	fields := map[string]interface{}{"value": value}
	return Write(influxClient, name, tags, fields, timestamp)
}

//WriteFloat64 takes a float64 value and sends it to influx
func WriteFloat64(influxClient client.Client, value float64, name string, tags map[string]string, timestamp time.Time) error {
	fields := map[string]interface{}{"value": value}
	return Write(influxClient, name, tags, fields, timestamp)
}

//Write sends data to influx
func Write(influxClient client.Client, name string, tags map[string]string, fields map[string]interface{}, timestamp time.Time) error {
	bp, err := client.NewBatchPoints(client.BatchPointsConfig{
		Database: "kubernetes",
	})
	if err != nil {
		return err
	}

	pt, err := client.NewPoint(name, tags, fields, timestamp)
	if err != nil {
		return err
	}
	bp.AddPoint(pt)

	err = influxClient.Write(bp)
	if err != nil {
		return err
	}
	return nil
}

//Connect to influx
func Connect() (client.Client, error) {
	influxClient, influxError := client.NewHTTPClient(client.HTTPConfig{
		Timeout: time.Second * 5,
		Addr:    fmt.Sprintf("http://%s", os.Getenv("DEIS_MONITOR_INFLUXAPI_SERVICE_HOST")),
	})
	if influxError != nil {
		return nil, influxError
	}
	return influxClient, nil
}


syslogish/server.go
package syslogish

import (
	"fmt"
	"log"
	"net"
	"strings"

	"github.com/deis/stdout-metrics/influx"
	"github.com/deis/stdout-metrics/util"
	"github.com/influxdata/influxdb/client/v2"
)

const (
	bindHost  = "0.0.0.0"
	bindPort  = 1514
	queueSize = 50000
)

// Server implements a UDP-based "syslog-like" server.  Like syslog, as described by RFC 3164, it
// expects that each packet contains a single log message and that, conversely, log messages are
// encapsulated in their entirety by a single packet, however, no attempt is made to parse the
// messages received or validate that they conform to the specification.
type Server struct {
	Conn         net.PacketConn
	Listening    bool
	StorageQueue chan string
	InfluxClient client.Client
}

// NewServer returns a pointer to a new Server instance.
func NewServer() (*Server, error) {
	addr, err := net.ResolveUDPAddr("udp", fmt.Sprintf("%s:%d", bindHost, bindPort))
	if err != nil {
		return nil, err
	}
	c, err := net.ListenUDP("udp", addr)
	if err != nil {
		return nil, err
	}

	influxClient, err := influx.Connect()
	if err != nil {
		return nil, err
	}

	return &Server{
		Conn:         c,
		StorageQueue: make(chan string, queueSize),
		InfluxClient: influxClient,
	}, nil
}

// Listen starts the server's main loop.
func (s *Server) Listen() {
	// Should only ever be called once
	if !s.Listening {
		defer s.InfluxClient.Close()
		s.Listening = true
		go s.receive()
		go s.parse()
		log.Println("syslogish server running")
	}
}

func (s *Server) receive() {
	// Make buffer the same size as the max for a UDP packet
	buf := make([]byte, 65535)
	for {
		n, _, err := s.Conn.ReadFrom(buf)
		if err != nil {
			log.Fatal("syslogish server read error", err)
		}
		message := strings.TrimSuffix(string(buf[:n]), "\n")
		select {
		case s.StorageQueue <- message:
		default:
		}
	}
}

func (s *Server) parse() {
	for message := range s.StorageQueue {
		messageJSON, err := util.ParseMessage(message)
		if err == nil && util.FromContainer(messageJSON, "deis-router") {
			err = influx.WriteRouterMetrics(s.InfluxClient, messageJSON)
			if err != nil {
				fmt.Printf("Error@Metrics:%v\n", err)
			}
		}
	}
}


influx/router_metrics.go
package influx

import (
	"time"

	"github.com/deis/stdout-metrics/util"
	"github.com/influxdata/influxdb/client/v2"
)

//WriteRouterMetrics converts json to influx datatype and sends
func WriteRouterMetrics(influxClient client.Client, json map[string]interface{}) error {
	parsedMessage, err := util.ParseNginxLog(json["log"].(string))
	if err != nil {
		return err
	}
	err = writeData(influxClient, util.GetHost(json), parsedMessage)
	if err != nil {
		return err
	}
	return nil
}

func writeData(influxClient client.Client, host string, parsedMessage map[string]interface{}) error {
	tags := map[string]string{"host": host,
		"app":         parsedMessage["app"].(string),
		"status_code": parsedMessage["status_code"].(string),
	}

	// We cant use the nginx timestamp because its not precise enough
	//parsedMessage["time"].(time.Time)
	err := WriteFloat64(influxClient, parsedMessage["request_time"].(float64), "deis_router_response_time_seconds", tags, time.Now())
	if err != nil {
		return err
	}
	return nil
}


influx/queries_per_second.go
package influx

import (
	"encoding/json"
	"fmt"
	"time"

	"github.com/influxdata/influxdb/client/v2"
)

var (
	queriesPerSecondAggregate = `SELECT count(value) FROM deis_router_response_time_seconds WHERE time > now() - 1m GROUP BY time(1m)`
	queriesPerSecondPerApp    = `SELECT count(value) FROM deis_router_response_time_seconds WHERE time > now() - 1m AND app = '%s' GROUP BY time(1m)`
	appQuery                  = `SHOW TAG VALUES FROM deis_router_response_time_seconds WITH KEY = app`
)

// This code isnt being used right now since we can perform the calculations in grafana. However, we are keeping this around just in case we need it in a future release.

//QueriesPerSecond finds the qps value for the aggregate and individual endpoints
func QueriesPerSecond(influxClient client.Client) error {
	err := writeAggregateQueriesPerSecond(influxClient)
	if err != nil {
		return err
	}
	err = writePerAppQueriesPerSecond(influxClient)
	if err != nil {
		return err
	}
	return nil
}

func writeAggregateQueriesPerSecond(influxClient client.Client) error {
	results, err := Query(influxClient, queriesPerSecondAggregate)
	if err != nil {
		return err
	}
	tags := map[string]string{"app": "aggregate"}
	err = writeQueriesPerSecond(influxClient, results, tags)
	if err != nil {
		return err
	}
	return nil
}

func writePerAppQueriesPerSecond(influxClient client.Client) error {
	results, err := Query(influxClient, appQuery)
	if err != nil {
		return err
	}

	if len(results[0].Series) > 0 && len(results[0].Series[0].Values) > 0 {
		for _, value := range results[0].Series[0].Values {
			app := value[1].(string)
			appResults, err := Query(influxClient, fmt.Sprintf(queriesPerSecondPerApp, app))
			if err != nil {
				return err
			}
			tags := map[string]string{"app": app}
			err = writeQueriesPerSecond(influxClient, appResults, tags)
			if err != nil {
				return err
			}
		}
	}
	return nil
}

func writeQueriesPerSecond(influxClient client.Client, results []client.Result, tags map[string]string) error {
	if len(results[0].Series) > 0 {
		v, _ := results[0].Series[0].Values[1][1].(json.Number).Float64()
		err := WriteFloat64(influxClient, v/60, "deis_router_request_count_seconds", tags, time.Now())
		if err != nil {
			return err
		}
	}
	return nil
}


_scripts/generate-changelog.sh
#!/usr/bin/env bash
# vim: et tw=0 sw=4 ts=4

REPOROOT=${REPOROOT:=https://github.com/deisthree/logger}
TYPES=${TYPES:=feat(;Features fix(;Fixes docs(;Documentation chore(;Maintenance}

usage() {
    echo "Usage: $0 <from> [to]

The REPOROOT and TYPES variables may be used to customize what links are
generated and what is scraped from the commit logs. Current settings are:

    REPOROOT=$REPOROOT
    TYPES=$TYPES
"
}

# Print the information scraped from git
retrieve() {
    while read -r commit hash message; do
        # Extract the subsystem where type(subsystem): message
        SUBSYSTEM=$(echo "$message" | cut -d'(' -f2 | cut -d')' -f1 | sed 's/\*/\(all\)/g')
        # Extract the message where type(subsystem): message
        MESSAGE=$(echo "$message" | awk -F ")" '{ print $2}' | sed 's/://' | cut -f2- -d' ')
        # Generate a link to the full legal commit on GitHub
        LINK="$REPOROOT/commit/$hash"
        # Echo all this in a way that makes the commit hash and message a link
        # to the commit in markdown
        echo ' -' "[\`$commit\`]($LINK)" "$SUBSYSTEM": "$MESSAGE"
    done < <(git --no-pager log --oneline --merges --oneline --format="%h %H %b" --grep="$1" "$FROM".."$TO")
    # Scrape the information from git
}

# Wrap feature type and show its relevant commits
subheading() {
    echo "#### $1"
    echo
    retrieve "$2"
    echo
}

main() {
    # Print usage summary if user didn't specify a beginning
    if [ -z "$1" ]; then
        usage
        exit 1
    fi

    FROM=$1
    TO=${2:-"HEAD"}

    printf "### %s -> %s\n\n" "$FROM" "$TO"

    # Iterate over the types of messages specified
    for LEGALTYPE in $TYPES; do
        SHORT=$(echo "$LEGALTYPE" | cut -f1 -d';')
        LONG=$(echo "$LEGALTYPE" | cut -f2 -d';')
        subheading "$LONG" "$SHORT"
    done
}

if [ $SHLVL -eq 2 ]; then
    # If this is being run as a command
    main "$@"
else
    # Otherwise this is being sourced
    unset -f main
    unset -f usage
fi


util/util.go
package util

import (
	"encoding/json"
	"fmt"
	"regexp"
	"strconv"
	"strings"
	"time"
)

var (
	//[2016-05-06T20:04:23+00:00]
	longForm = "[2006-01-02T15:04:05-07:00]"
)

func fromKubernetes(json map[string]interface{}) bool {
	return json["kubernetes"] != nil
}

// FromContainer checks that the log message is from a given container
func FromContainer(json map[string]interface{}, pattern string) bool {
	if !fromKubernetes(json) {
		return false
	}

	containerName := json["kubernetes"].(map[string]interface{})["container_name"].(string)
	matched, _ := regexp.MatchString(pattern, containerName)
	return matched
}

//ParseMessage takes a string and returns a map
func ParseMessage(message string) (map[string]interface{}, error) {
	curlyIndex := strings.Index(message, "{")
	if curlyIndex > -1 {
		message = message[curlyIndex:]
		var messageJSON map[string]interface{}
		err := json.Unmarshal([]byte(message), &messageJSON)
		if err == nil {
			return messageJSON, nil
		}
		return nil, err
	}
	return nil, fmt.Errorf("Not a valid JSON message: %s", message)
}

//GetHost returns the host value from the kubernetes submap
func GetHost(json map[string]interface{}) string {
	return json["kubernetes"].(map[string]interface{})["host"].(string)
}

//ParseNginxLog returns map of the parsed nginx log
func ParseNginxLog(message string) (map[string]interface{}, error) {
	parsedMessage := make(map[string]interface{})
	splitMessage := strings.Split(message, " - ")
	if len(splitMessage) > 1 {
		timestamp, err := toTime(splitMessage[0])
		if err != nil {
			return nil, err
		}

		bytesSent, err := strconv.Atoi(strings.TrimSpace(splitMessage[5]))
		if err != nil {
			return nil, err
		}

		responseTime, err := strconv.ParseFloat(strings.TrimSpace(splitMessage[11]), 64)
		if err != nil {
			return nil, err
		}

		requestTime, err := strconv.ParseFloat(strings.TrimSpace(splitMessage[12]), 64)
		if err != nil {
			return nil, err
		}

		parsedMessage["time"] = timestamp
		parsedMessage["app"] = strings.TrimSpace(splitMessage[1])
		parsedMessage["status_code"] = strings.TrimSpace(splitMessage[3])
		parsedMessage["bytes_sent"] = bytesSent
		parsedMessage["response_time"] = responseTime
		parsedMessage["request_time"] = requestTime
		return parsedMessage, nil
	}
	return nil, fmt.Errorf("Invalid nginx log message: %s", message)
}

//Turn this [2016-05-06T20:04:23+00:00] into time object
func toTime(date string) (time.Time, error) {
	t, err := time.Parse(longForm, date)
	if err != nil {
		return time.Now(), err
	}
	return t, nil
}
